### PR TITLE
profitbricks: support for update state, check mode, custom API URL & other improvements

### PR DESCRIPTION
**Pull request by nurfet-becirevic**
##### SUMMARY
This update contains new features and other changes including:

- Added support for update state. The following ProfitBricks components can be updated now: datacenters, servers,  NICs and volumes.
- Added support for check mode.
- Added support for custom ProfitBricks  API URL.
- Deprecated redundant module `profitbricks_volume_attachments`. The module functionality is handled by `profitbricks` module.
- Added `profitbricks` module to `module_utils` grouping common stuff from ProfitBricks  cloud modules.
- Added ProfitBricks credential parameters `username` & `password` replacing `subscription_user` and `subscription_password` which are defined as aliases now.
- Other ProfitBricks related updates: support for API v4, request headers info, missing parameters in the modules etc.
- Fixed/updated documentation.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
various ProfitBricks modules

##### ANSIBLE VERSION
```
2.6
```

##### ADDITIONAL INFORMATION
```
Several new ProfitBricks modules and dynamic inventory are waiting in the queue for PR. They have been omitted here to facilitate the review.  
```

### CODE REVIEW

**Comment by nurfet-becirevic** [comment]
I am not sure how to handle `ImportError: cannot import name __version__`. None of these seems to work:

```from ansible import __version__```
```from ansible import __version__ as ansible_version```
```from ansible.release import __version__```

**Comment by sivel** [comment]
Instead of trying to import `__version__`, instead utilize `module.ansible_version`.

**Comment by nurfet-becirevic** [comment]
```
2018-01-15 21:11:59 ERROR: Command "python3.6 test/sanity/validate-modules/validate-modules --format json --arg-spec lib/ansible/modules/cloud/profitbricks/profitbricks.py lib/ansible/modules/cloud/profitbricks/profitbricks_datacenter.py lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py lib/ansible/modules/cloud/profitbricks/profitbricks_volume.py --exclude '^(lib/ansible/modules/utilities/logic/async_status.py|lib/ansible/modules/utilities/helper/_accelerate.py)' --base-branch origin/devel" returned exit status 1.
2018-01-15 21:11:59 >>> Standard Error
2018-01-15 21:11:59 Traceback (most recent call last):
2018-01-15 21:11:59   File "test/sanity/validate-modules/validate-modules", line 1361, in <module>
2018-01-15 21:11:59     main()
2018-01-15 21:11:59   File "test/sanity/validate-modules/validate-modules", line 1260, in main
2018-01-15 21:11:59     mv.validate()
2018-01-15 21:11:59   File "test/sanity/validate-modules/validate-modules", line 1152, in validate
2018-01-15 21:11:59     doc_info = self._validate_docs()
2018-01-15 21:11:59   File "test/sanity/validate-modules/validate-modules", line 876, in _validate_docs
2018-01-15 21:11:59     self._check_for_new_args(doc)
2018-01-15 21:11:59   File "test/sanity/validate-modules/validate-modules", line 1067, in _check_for_new_args
2018-01-15 21:11:59     names = [option] + details.get('aliases', [])
2018-01-15 21:11:59 TypeError: can only concatenate list (not "str") to list
```

Why the CI build is complaining on python 3.6 compatibility? Shouldn't it fail on the following as well?
https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/system/user.py#L2119-L2120

**Comment by sivel** [comment]
The problem you are experiencing is that `aliases` needs to be a list, and you have defined it as a string:

https://github.com/ansible/ansible/pull/34834/files#diff-01759fa310331bba471354ce4323085cR133

Should be:

```diff
+  username:
+    description:
+      - The ProfitBricks username. Overrides the PROFITBRICKS_USERNAME environment variable.
+    required: false
+    aliases:
+      - subscription_user
```

An update likely needs to be made to validate-modules to handle this more gracefully

**Comment by nurfet-becirevic** [comment]
bot_status

**Comment by edevenport** [comment]
shipit

**Comment by baldwinSPC** [comment]
shipit

**Comment by edevenport** [comment]
@nurfet-becirevic Can you address this conflict so we can have a clean merge?

**Comment by nurfet-becirevic** [comment]
@edevenport It is resolved

**Comment by edevenport** [comment]
@resmo @mattclay Is there anything we can do to move this PR forward?

File: lib/ansible/module_utils/profitbricks.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Copyright: Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```
Comment (resmo) [inline lib/ansible/module_utils/profitbricks.py]: code in module_utils must be BSD licensed

**Review by resmo — DISMISSED** [review]

File: lib/ansible/module_utils/profitbricks.py (hunk 1 above)
Comment (nurfet-becirevic) [reply lib/ansible/module_utils/profitbricks.py]: @resmo It's fixed

**Comment by resmo** [comment]
Note that modules once in ansible can not just be deleted when they become obsolete. This would break users playbooks. Modules must be deprecated. See http://docs.ansible.com/ansible/devel/dev_guide/developing_modules_checklist.html#deprecating-and-making-module-aliases

**Comment by nurfet-becirevic** [comment]
No idea why the build failed. It doesn't seem to be related to my changes. Rebasing to trigger a new build.

**Comment by nurfet-becirevic** [comment]
bot_status

**Comment by nurfet-becirevic** [comment]
Is there any additional change request, @resmo ?

**Comment by edevenport** [comment]
@resmo Just checking in again to see if any further changes are necessary for this PR?

File: test/sanity/validate-modules/ignore.txt [hunk 2]
```diff
@@ -506,10 +506,10 @@ lib/ansible/modules/cloud/profitbricks/profitbricks.py E324
 lib/ansible/modules/cloud/profitbricks/profitbricks.py E326
 lib/ansible/modules/cloud/profitbricks/profitbricks_datacenter.py E326
 lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py E324
+lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py E325
```
Comment (mattclay) [inline test/sanity/validate-modules/ignore.txt]: New ignores should not be added. The underlying issue should be fixed instead. If the issue is in common code (module_utils or docs fragments), those issues should be fixed in a separate PR first.

Comment (nurfet-becirevic) [reply test/sanity/validate-modules/ignore.txt]: Resolved, thanks @mattclay

**Comment by edevenport** [comment]
@nurfet-becirevic Can you fix the current conflict and ping the upstream maintainer again?

**Comment by nurfet-becirevic** [comment]
The change log has been removed in the upstream. Is there another way to document deprecated modules, @mattclay ?

**Comment by mattclay** [comment]
@nurfet-becirevic Add a changelog fragment: 

https://github.com/ansible/ansible/tree/devel/changelogs/fragments

See the existing fragments for examples.

**Comment by nurfet-becirevic** [comment]
@mattclay It's done. Please let us know if any other change is required for this PR.

**Comment by edevenport** [comment]
@resmo Can you review/approve the changes? @mattclay Is there anything else we need to get this merged?

File: lib/ansible/modules/cloud/profitbricks/_profitbricks_volume_attachments.py [hunk 3]
```diff
@@ -31,20 +35,29 @@
     description:
       - The volume name or ID.
     required: true
-  subscription_user:
+  api_url:
     description:
-      - The ProfitBricks username. Overrides the PB_SUBSCRIPTION_ID environment variable.
+      - The ProfitBricks API base URL.
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/_profitbricks_volume_attachments.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/_profitbricks_volume_attachments.py [hunk 4]
```diff
@@ -18,6 +18,10 @@
 description:
      - Allows you to attach or detach a volume from a ProfitBricks server. This module has a dependency on profitbricks >= 1.0.0
 version_added: "2.0"
+deprecated:
+    removed_in: "2.10"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/_profitbricks_volume_attachments.py]: `"2.12"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks.py [hunk 5]
```diff
@@ -10,95 +10,140 @@
...
       - Public SSH keys allowing access to the virtual machine.
-    version_added: '2.2'
+    required: false
+    version_added: "2.2"
+  volume_availability_zone:
+    description:
+      - The storage availability zone assigned to the volume.
+    default: null
+    choices: [ "AUTO", "ZONE_1", "ZONE_2", "ZONE_3" ]
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks.py [hunk 6]
```diff
@@ -10,95 +10,140 @@
...
     choices: [ "AMD_OPTERON", "INTEL_XEON" ]
-    version_added: '2.2'
+    version_added: "2.2"
+  availability_zone:
+    description:
+      - The availability zone assigned to the server.
+    required: false
+    default: AUTO
+    choices: [ "AUTO", "ZONE_1", "ZONE_2" ]
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks.py [hunk 7]
```diff
@@ -10,95 +10,140 @@
...
+    required: false
     default: 1
-  subscription_user:
+  nat:
+    description:
+      - Boolean value indicating if the private IP address has outbound access to the public Internet.
+    type: bool
+    required: false
+    default: false
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks.py [hunk 8]
```diff
@@ -10,95 +10,140 @@
...
+  api_url:
+    description:
+      - The ProfitBricks API base URL.
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks.py [hunk 9]
```diff
@@ -133,11 +180,24 @@
     ram: 2048
     volume_size: 50
     cpu_family: INTEL_XEON
-    image: a3eae284-a2fe-11e4-b187-5f1f641608c8
+    image: ubuntu:latest
     location: us/las
     count: 3
     assign_public_ip: true
 
+# Update Virtual machines
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks.py]: Can you please use `- name: ` rather than `# comments`

File: lib/ansible/modules/cloud/profitbricks/profitbricks.py [hunk 10]
```diff
@@ -133,11 +180,24 @@
...
+
+- profitbricks:
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks.py]: remember to remove the `-` from here when you add `- name:`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py [hunk 11]
```diff
@@ -34,21 +33,53 @@
...
       - The LAN to place the NIC on. You can pass a LAN that doesn't exist and it will be created. Required on create.
-    required: true
-  subscription_user:
+    default: null
+  nat:
+    description:
+      - Boolean value indicating if the private IP address has outbound access to the public internet.
+    type: bool
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py [hunk 12]
```diff
@@ -34,21 +33,53 @@
...
+  dhcp:
+    description:
+      - Boolean value indicating if the NIC is using DHCP or not.
+    type: bool
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py [hunk 13]
```diff
@@ -34,21 +33,53 @@
...
+  firewall_active:
     description:
-      - The ProfitBricks username. Overrides the PB_SUBSCRIPTION_ID environment variable.
+      - Boolean value indicating if the firewall is active.
+    type: bool
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py [hunk 14]
```diff
@@ -34,21 +33,53 @@
...
+  ips:
+    description:
+      - A list of IPs to be assigned to the NIC.
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py [hunk 15]
```diff
@@ -34,21 +33,53 @@
...
+  api_url:
+    description:
+      - The ProfitBricks API base URL.
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_nic.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_volume.py [hunk 16]
```diff
@@ -63,7 +62,13 @@
...
     required: false
     default: UNKNOWN
-    choices: ["LINUX", "WINDOWS", "UNKNOWN" , "OTHER"]
+    choices: ["LINUX", "WINDOWS", "UNKNOWN" , "OTHER", "WINDOWS2016"]
+  availability_zone:
+    description:
+      - The storage availability zone assigned to the volume.
+    default: null
+    choices: [ "AUTO", "ZONE_1", "ZONE_2", "ZONE_3" ]
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_volume.py]: `"2.8"`

File: lib/ansible/modules/cloud/profitbricks/profitbricks_volume.py [hunk 17]
```diff
@@ -72,39 +77,52 @@
...
     description:
       - list of instance ids, currently only used when state='absent' to remove instances.
     required: false
-  subscription_user:
+  api_url:
     description:
-      - The ProfitBricks username. Overrides the PB_SUBSCRIPTION_ID environment variable.
+      - The ProfitBricks API base URL.
+    default: null
+    version_added: "2.6"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/profitbricks/profitbricks_volume.py]: `"2.8"`

File: lib/ansible/module_utils/profitbricks.py [hunk 18]
```diff
@@ -0,0 +1,105 @@
+# This code is part of Ansible, but is an independent component.
```
Comment (gundalow) [inline lib/ansible/module_utils/profitbricks.py:1]: Please moved to `module_utils/cloud/profitbricks/profitbricks.py` you will need to create `__init__.py` files in any new directories.

Please also add this new directory to https://github.com/ansible/ansible/blob/devel/.github/BOTMETA.yml

**Review by gundalow — COMMENTED** [review]
If there are any common options they could go into `module_utils` rather than being repeated in each module. The docs should go into `lib/ansible/utils/module_docs_fragments` see  https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_in_groups.html#your-first-pull-request

That should help to reduce the side of the modules a bit

File: lib/ansible/modules/cloud/profitbricks/_profitbricks_volume_attachments.py (hunk 4 above)
Comment (nurfet-becirevic) [reply lib/ansible/modules/cloud/profitbricks/_profitbricks_volume_attachments.py]: it seems that a `validate-modules` test complains on `2.12` here.

**Comment by gundalow** [comment]
Ah, once https://github.com/ansible/ansible/pull/45732 is merged `"2.12"` will be allowed

**Comment by gundalow** [comment]
@nurfet-becirevic CI is now green

File: lib/ansible/module_utils/profitbricks.py (hunk 18 above)
Comment (nurfet-becirevic) [reply lib/ansible/module_utils/profitbricks.py:1]: @gundalow , just to make sure before I commit the changes, the parent dir `module_utils/cloud` does not exist either. `profitbricks` is the first util module that is to be residing under `module_utils/cloud` , right?

Comment (gundalow) [reply lib/ansible/module_utils/profitbricks.py:1]: Correct

**Comment by nurfet-becirevic** [comment]
I have applied the requested changes @gundalow . CI failure doesn't seem to be related to my latest commit.

**Comment by nurfet-becirevic** [comment]
bot_status

File: lib/ansible/module_utils/cloud/profitbricks/profitbricks.py [hunk 19]
```diff
@@ -0,0 +1,105 @@
+# This code is part of Ansible, but is an independent component.
```
Comment (mattclay) [inline lib/ansible/module_utils/cloud/profitbricks/profitbricks.py]: Put this in `lib/ansible/module_utils/profitbricks.py` instead.

**Comment by mattclay** [comment]
The CI failure is due to the conflict between:

* `lib/ansible/module_utils/cloud.py`
* `lib/ansible/module_utils/cloud/`

Using `lib/ansible/module_utils/profitbricks.py` should resolve that.

File: lib/ansible/module_utils/cloud/profitbricks/profitbricks.py (hunk 19 above)
Comment (nurfet-becirevic) [reply lib/ansible/module_utils/cloud/profitbricks/profitbricks.py]: @mattclay, basically, you are saying I should revert the last commit? Before my last commit `profitbricks.py` was in `lib/ansible/module_utils/`.

Comment (mattclay) [reply lib/ansible/module_utils/cloud/profitbricks/profitbricks.py]: Yes, although I didn't see the comment from @gundalow earlier that requested the change. We'll need to do something to avoid the conflict. Reverting the change would be the easiest.

**Comment by samdoran** [comment]
!waffling ci_verified

**Comment by gundalow** [comment]
bot_status

**Comment by Akasurde** [comment]
Please submit a new module request to https://github.com/ansible-collections/community.general. Thanks for your contribution.
